portal-impl/src/com/liferay/portlet/internal/PortletURLImpl.java

```diff
@@ -960,21 +1003,125 @@ protected String generateToString() {
 
-			renderParams = _mergeWithRenderParameters(renderParams);
+			for (String parameterName : resourceParameterNames) {
+				portletURLParams.put(
+					_RESOURCE_PARAMETER_NAMESPACE + parameterName,
+					_mutableResourceParameters.getValues(parameterName));
+			}
 		}
 
-		for (Map.Entry<String, String[]> entry : renderParams.entrySet()) {
-			String name = entry.getKey();
+		if (!_lifecycle.equals(PortletRequest.RESOURCE_PHASE) ||
+			(_lifecycle.equals(PortletRequest.RESOURCE_PHASE) &&
+			 _copyCurrentRenderParameters &&
+			 !_cacheability.equals(ResourceURL.FULL))) {
 
-			if (isParameterIncludedInPath(name)) {
-				continue;
+			Set<String> renderParameterNames =
+				_mutableRenderParameters.getNames();
+
+			for (String renderParameterName : renderParameterNames) {
+				if (!resourceParameterNames.contains(renderParameterName)) {
+					if (_lifecycle.equals(PortletRequest.RESOURCE_PHASE) &&
+						_mutableRenderParameters.isPublic(
+							renderParameterName)) {
+
+						continue;
+					}
+
+					if (!_lifecycle.equals(PortletRequest.RESOURCE_PHASE) &&
+						(_removedParameterNames != null) &&
+						_removedParameterNames.contains(renderParameterName)) {
+
+						continue;
+					}
+
+					String[] renderParameterValues =
+						_mutableRenderParameters.getValues(renderParameterName);
+
+					if (_mutableRenderParameters.isPublic(
+							renderParameterName)) {
+
+						portletURLParams.put(
+							renderParameterName, renderParameterValues);
+
+						continue;
+					}
+
+					if (_lifecycle.equals(PortletRequest.ACTION_PHASE) &&
+						actionParameterNames.contains(renderParameterName)) {
+
+						String[] actionParameterValues =
+							_mutableActionParameters.getValues(
+								renderParameterName);
+
+						if ((actionParameterValues != null) &&
+							_copyCurrentRenderParameters) {
+
+							renderParameterValues = ArrayUtil.append(
+								actionParameterValues, renderParameterValues);
+						}
+
+						renderParameterName =
+							_ACTION_PARAMETER_NAMESPACE + renderParameterName;
+					}
+					else if (_lifecycle.equals(PortletRequest.RENDER_PHASE)) {
+						PortletRequest portletRequest = getPortletRequest();
+
+						if (portletRequest != null) {
+							LiferayRenderParameters renderParameters =
+								(LiferayRenderParameters)
+									portletRequest.getRenderParameters();
+
+							String[] requestRenderParameterValues =
+								renderParameters.getValues(renderParameterName);
+
+							if ((requestRenderParameterValues != null) &&
+								_copyCurrentRenderParameters &&
+								!Arrays.equals(
+									requestRenderParameterValues,
+									renderParameterValues)) {
+
+								renderParameterValues = ArrayUtil.append(
+									renderParameterValues,
+									requestRenderParameterValues);
+							}
+						}
+					}
+					else {
+						renderParameterName =
+							PortletQName.PRIVATE_RENDER_PARAMETER_NAMESPACE +
+								renderParameterName;
```

Comment:
@shuyangzhou Sorry -- I should have called `String.concat(String)` here. Perhaps you could fix that in a code review commit.

Regardless, it is an important line for me to direct your attention to, because it points out the new `priv_r_p_` namespace prefix introduced by the new `PortletQName.PRIVATE_RENDER_PARAMETER_NAMESPACE` constant that I added a while back.

Here is the problem that we needed to solve -- given the following JSP code:
```
<portlet:actionURL>
   <param name="foo" value="1234" />
   <param name="bar" value="5678" type="render" />
</portlet:actionURL>
```
How do you make a distinction between the *action parameter* named `"foo"` and the *private render parameter* named `"bar"? Also, how do you do it in such a way that you don't rewrite the entire URL generation scheme of Liferay Portal's portlet container?

The answer was to add the new `"priv_r_p_"` namespace prefix for private render parameters. It was the least-invasive thing I could do in order to make the distinction between action parameters and private render parameters in the same PortletURL.